qtfred/src/mission/EditorViewport.cpp

```diff
@@ -318,9 +319,22 @@ void EditorViewport::process_system_keys(int key) {
 }
 
 void EditorViewport::process_controls(vec3d* pos, matrix* orient, float frametime, int key, int mode) {
+	static std::unique_ptr<io::spacemouse::SpaceMouse> spacemouse = io::spacemouse::SpaceMouse::searchSpaceMouses(0);
+
 	if (Flying_controls_mode) {
 		grid_read_camera_controls(&view_controls, frametime);
 
+		if (spacemouse != nullptr) {
+			auto spacemouse_movement = spacemouse->getMovement();
+			spacemouse_movement.handleNonlinearities(Fred_spacemouse_nonlinearity);
+			view_controls.pitch += spacemouse_movement.rotation.p;
+			view_controls.vertical += spacemouse_movement.translation.xyz.z;
+			view_controls.heading += spacemouse_movement.rotation.h;
+			view_controls.sideways += spacemouse_movement.translation.xyz.x;
+			view_controls.bank += spacemouse_movement.rotation.b;
+			view_controls.forward += spacemouse_movement.translation.xyz.y;
```

Comment:
Yeah.
While I haven't tested it in QTFred but just normal FRED (I assume that they react identically to the axis mapping), these axis mappings are so that it feels intuitive to me.
The actual x/y/z mappings of the translation vector are chosen in such a way that for the other mode that uses not the CI-struct but actual vec3d / angles, I can just add the vector and not have to change around axes.